ARMmbed/mbed-os
# Initial adding LLVM 6.0 Toolchain

Hi ARMmbed developers,

## Description

Build ARMmbed [mbed-os-example-wifi](https://github.com/ARMmbed/mbed-os-example-wifi) for [REALTEK_RTL8195AM](https://os.mbed.com/platforms/Realtek-RTL8195AM/) target via LLVM Toolchain based on clang compiler, lld linker, llvm-objcopy, musl libc, llvm's libcxx and compiler-rt.

## Status

**IN DEVELOPMENT**

## Todos

- [ ] Link with `lib_peripheral_mbed_llvm.a` and `lib_wlan_mbed_llvm.a`
- [ ] Introduce how to use LLVM Toolchain to build wlan library for targets

## Steps to test or reproduce

Follow the "Building Clang and Working with the Code" section in [Getting Started: Building and Running Clang](http://clang.llvm.org/get_started.html), after step 3, it also needs to check out lld:

```
cd llvm/tools
svn co http://llvm.org/svn/llvm-project/lld/trunk lld
```

cmake build option tested for Fedora 25:

```
cmake .. -DCMAKE_BUILD_TYPE=Release \
    -DCMAKE_INSTALL_PREFIX=/opt/llvm-svn \
    -DCMAKE_C_COMPILER=clang \
    -DCMAKE_CXX_COMPILER=clang++ \
    -DCMAKE_C_FLAGS="-fPIC" \
    -DCMAKE_CXX_FLAGS="-std=c++11 -fPIC" \
    -DLLVM_BUILD_LLVM_DYLIB=ON \
    -DLLVM_LINK_LLVM_DYLIB=ON \
    -DLLVM_INSTALL_UTILS=ON \
    -DLLVM_ENABLE_RTTI=ON \
    -DLLVM_ENABLE_FFI=ON \
    -DLLVM_ENABLE_EH=ON \
    -DLLVM_BUILD_TESTS=ON \
    -DLLVM_BUILD_DOCS=OFF \
    -DLLVM_ENABLE_SPHINX=OFF \
    -DLLVM_ENABLE_DOXYGEN=OFF \
    -DLLDB_DISABLE_LIBEDIT=1 \
    -DSPHINX_WARNINGS_AS_ERRORS=OFF \
    -DFFI_INCLUDE_DIR=$(pkg-config --variable=includedir libffi) \
    -DFFI_LIBRARY_DIR:PATH="$(pkg-config --variable=libdir libffi)" \
    -DLLVM_BINUTILS_INCDIR=/usr/include \
    -DLLVM_LIBDIR_SUFFIX=64 \
    -DLIBCXX_ENABLE_ABI_LINKER_SCRIPT=ON \
    -DLIBUNWIND_ENABLE_SHARED=ON \
    -DLIBCXXABI_USE_LLVM_UNWINDER=ON \
    -DLLVM_BUILD_EXAMPLES=OFF \
    -DLLDB_TEST_C_COMPILER=clang \
    -DLLDB_TEST_CXX_COMPILER=clang++ \
    -DLLVM_DEFAULT_TARGET_TRIPLE="x86_64-redhat-linux" \
    -DCLANG_VENDOR="Fedora"
```

configure build option for musl libc:

```
LLVM_SVN_BIN=/opt/llvm-svn/bin

export PATH=$PATH:$LLVM_SVN_BIN
export CC=$LLVM_SVN_BIN"/clang --target=thumbv7em-unknown-none-eabi -mthumb -march=armv7e-m -mcpu=cortex-m3 -mfloat-abi=soft -mfpu=fpv5-sp-d16 -ffreestanding -mlittle-endian -mimplicit-it=always"
export LD=$LLVM_SVN_BIN"/ld.lld"

./configure --build=arm-none-eabi \
    --host=x86_64-redhat-gnu \
    --target=arm-none-eabi \
    --prefix=/opt/arm-musl \
    --disable-shared
```

Also it is able to [download ellcc libraries](http://ellcc.org/releases/latest/ellcc-libraries-2017-08-23.bz2) including musl, libcxx and compiler-rt, etc. for armv7e-m.


Copy the [Makefile](https://github.com/xiangzhai/ufiasco/blob/llvm-toolchain/Makefile) template and run `make` to build.

Regards,
Leslie Zhai - a LLVM developer https://reviews.llvm.org/p/xiangzhai/
